### PR TITLE
add between/3

### PR DESCRIPTION
**Pull request by guregu**
This adds between/3 as discussed in #163.

Thanks for the advice about control flow! I tried refactoring it.
Let me know if I should make any changes and I will be happy to.

BTW, should I add this to the default predicates too?

### CODE REVIEW

File: engine/builtin.go [hunk 1]
```diff
@@ -813,6 +813,50 @@ func Compare(order, term1, term2 Term, k func(*Env) *Promise, env *Env) *Promise
...
+		return Error(TypeErrorInteger(upper))
+	}
+
+	if low > high {
+		return Bool(false)
+	}
+
+	switch value := env.Resolve(value).(type) {
+	case Integer:
+		if value >= low && value <= high {
```
Comment (ichiban) [inline engine/builtin.go:845]: This is a minor point but it'd be better to flow to `k(env)`.

> The code reads well if the successful flow of control runs down the page, eliminating error cases as they arise.
https://go.dev/doc/effective_go#if

```
  *
  |
  +--> Bool(false)
  |
  +--> Error(err)
  |
  v
k(env)
```

File: engine/builtin.go [hunk 2]
```diff
@@ -813,6 +813,50 @@ func Compare(order, term1, term2 Term, k func(*Env) *Promise, env *Env) *Promise
...
+			return k(env)
+		}
+		return Bool(false)
+	case Variable:
+		return Delay(func(context.Context) *Promise {
+			return Unify(value, low, k, env)
+		}, func(context.Context) *Promise {
+			return Between(low+1, upper, value, k, env)
```
Comment (ichiban) [inline engine/builtin.go:853]: When `low = high = math.MaxInt64`, `low+1` overflows and becomes a negative integer.

This could be solved by checking if `low < high`:

```go
if low == high { // Since we've already checked low <= high.
	return Bool(false)
}
return Between(low+1, upper, value, k, env)
```

Or, for a better performance:

```go
ks := make([]func(*Env) *Promise, 0, 2)
ks = append(ks, func(context.Context) *Promise {
	return Unify(value, low, k, env)
})
if low < high {
	ks = append(ks, func(context.Context) *Promise {
		return Between(low+1, upper, value, k, env)
	})
}
return Delay(ks...)
```

**Review by ichiban — COMMENTED** [review]
AWESOME!! There might be a weird behaviour on the edge case of `low = high = math.MaxInt64` so could you check it, please?

Other than that, this is a superb contribution. Thanks!

**Comment by ichiban** [comment]
This should be definitely in the default predicates! You can register it anywhere in `New()` (totally unordered).

https://github.com/ichiban/prolog/blob/66a3f9f8dae65df9f649c94fbdfaf267742d059f/interpreter.go#L23

File: engine/builtin.go (hunk 1 above)
Comment (guregu) [reply engine/builtin.go]: Good point! I tweaked it a bit. Should we go further and put k(env) at the very end? I'm OK with either way.

File: engine/builtin.go (hunk 2 above)
Comment (guregu) [reply engine/builtin.go]: Thanks! I changed it to the better performance suggestion.
Before it would overflow but get caught by 
```go
if low > high {
	return Bool(false)
}
```
but now it should completely eliminate that dead end 👍

**Comment by guregu** [comment]
> This should be definitely in the default predicates! You can register it anywhere in `New()` (totally unordered).
> 
> https://github.com/ichiban/prolog/blob/66a3f9f8dae65df9f649c94fbdfaf267742d059f/interpreter.go#L23

Awesome! I added it below Compare in builtin.go and interpreter.go because it seems kind of similar to Compare 🙂

**Comment by guregu** [comment]
> There might be a weird behaviour on the edge case of `low = high = math.MaxInt64` so could you check it, please?

Added a couple extra tests using MaxInt64 👍 
Thanks for the review! I committed a few changes based on your comments.
Let me know if I missed something.

**Comment by guregu** [comment]
I added a few more tests. This should bring coverage up to 100% I think.
